### ISSUE TITLE
Add ACL features to (global) tags

### ISSUE THREAD

**Issue by noseshimself** [nextcloud/server#32454]
**Is your feature request related to a problem? Please describe.**
There are use cases that make access right management cumbersome (with the directory structure of file storage being a tree) hard without having per-file access control. A typical application requiring this would be any kind of document management system-like usage of shared folders where certain documents should not be visible to anyone outside a group (e. g. only accounting and the department who ordered the stuff seeing invoices in the "incoming" directory) .

**Describe the solution you'd like**
Revamping the entire core to have file access control on every file would probably become a nightmare and (see current papers on the topic) might not be necessary. A just as versatile model would be adding ACLs (only for groups if you want to simplify -- single users could be put in arbitrary groups containing only one user) to the existing (global) tags. Tagging a file as "accounting" no matter where will add or set the specified access rights to the file (but only if the directory is shared with a user who is member of that group).

**Describe alternatives you've considered**
Using different software. Putting files in separate directories and sharing each of these leaf nodes to different groups of accounts is completely impractical.

**Additional context**
It's the only feature missing to replace most DMS and integrate them into Nextcloud.

**Comment by joshtrichards** [comment]
Hi @noseshimself - Have you explored a solution utilizing Flows and/or `groupfolders`?

https://docs.nextcloud.com/server/latest/admin_manual/file_workflows/index.html

https://github.com/nextcloud/groupfolders

**Comment by noseshimself** [comment]
> Hi @noseshimself - Have you explored a solution utilizing Flows and/or `groupfolders`?

Yes but that won't really solve the problem I need a solution for and it is something that is really missing. Data can't always be move into common locations and creating arbitrary sub-folders is also impossible or at least impractical.

You are thinking of tags as properties and processing marks for workflows. But that's just a minor part of what I would need them for. Nextcloud is lacking  non-tree-structured directories as well as virtual folders. These are an integral part of modern methods of organizing data

Please keep #32455 in the back of your mind, too, which is the other (and just as important half) of this request.

The general idea is something akin to providing features of a DMS/AMS. It should be able to tag documents with non-global tags (i. e. tags that are private for a (group of) user(s)). Simple example: You want to tag all files relevant for your taxes (income, account statements, bills, insurances...) that are distributed across a random collection of directories (in modern models: create a virtual directory) for easy access to them as a topic. No workflows involved, just finding them. Now add an external accountant dealing with everything tax -- you do not want to share every single file using the rather clumsy user interface for this purpose (especially if you have to tell them how to find all the files). You could of course create copies (there are no symbolic links that would clumsily emulate the tags either...) but sometimes files do change so static copies will be useless, too.

The two requested features (this and #32455) will, together with workflows) provide all the support infrastructure to turn Nextcloud into a fully capable asset management system.

Besides: Did you ever see a Nextcloud system where 500 users are using (their own) tags for file management defining some thousands of tags? Awesomely awful even with rules for tag name creation.

My goal is a replacement of our current DMS. We have a scan-to-OCR workflow that will provide files ready for storage to be able to hand them over to Nextcloud workflows for storing the documents in appropriate places What's missing are the virtual folders and tags all AMS/DMS provide these days. The mechanisms are redundant; you usually use virtual folders for ACL management and tags for grouping. If you are able to use tags to carry ACLs you can do it without the virtual folder mechanism and merging in the implementation of this kind of ACL should be much simpler (call it "sharing with a tag") then heaping more mechanisms.
